src/test/crypto_tests.cpp

```diff
@@ -857,4 +858,58 @@ BOOST_AUTO_TEST_CASE(sha3_256_tests)
 }
 
+static MuHash3072 FromInt(unsigned char i) {
+    unsigned char tmp[32] = {i, 0};
+    return MuHash3072(tmp);
+}
+
+BOOST_AUTO_TEST_CASE(muhash_tests)
+{
+    unsigned char out[384];
+
+    for (int iter = 0; iter < 10; ++iter) {
+        unsigned char res[384];
+        int table[4];
+        for (int i = 0; i < 4; ++i) {
+            table[i] = g_insecure_rand_ctx.randbits(3);
+        }
+        for (int order = 0; order < 4; ++order) {
+            MuHash3072 acc;
+            for (int i = 0; i < 4; ++i) {
+                int t = table[i ^ order];
+                if (t & 4) {
+                    acc /= FromInt(t & 3);
+                } else {
+                    acc *= FromInt(t & 3);
+                }
+            }
+            acc.Finalize(out);
+            if (order == 0) {
+                memcpy(res, out, 384);
+            } else {
+                BOOST_CHECK(memcmp(res, out, 384) == 0);
+            }
+        }
+
+        MuHash3072 x = FromInt(g_insecure_rand_ctx.randbits(4)); // x=X
+        MuHash3072 y = FromInt(g_insecure_rand_ctx.randbits(4)); // x=X, y=Y
+        MuHash3072 z; // x=X, y=Y, z=1
+        z *= x; // x=X, y=Y, z=X
+        z /= y; // x=X, y=Y, z=X/Y
```

Comment:
The headers suggest that removing things that aren't in the set is a bad idea ("not clear if the security assumptions still hold", "should for now only be used to represent a set of elements") so seems a bit odd to be doing it in the tests. Could convert it to `z *= x; z *= y; y *= x; z /= y;` to only add things that aren't present/remove things that are present.

I think the comment in the header doesn't make sense though -- as I understand it, it should be equally fine (in security terms even) to remove a set by saying:

```
    MuHash3072 to_remove;
    for (x : spent) { to_remove *= MuHash3072(x); }
    set /= to_remove;
```

and

```
    MuHash3072 changes;
    for (x : arrived) { changes *= MuHash3072(x); }
    for (x : spent) { changes /= MuHash3072(x); }
    set *= changes;
```

even though that's removing things from `changes` before they've been added.

Reply:
Yeah, I think this makes sense. From what I have read I think there was just no research done on the security of this. I think it makes sense to be defensive but I amended the comment that this would just not the intended use. I also updated the test to follow this.